controllers/networkplugins_controller.go

```diff
@@ -17,35 +17,584 @@ limitations under the License.
+			if err != nil {
+				return err
+			}
+			*fileList = append(*fileList, "sriov-cni.yaml")
+			*fileList = append(*fileList, "sriov-deviceplugin.yaml")
+		}
+
+		if plugins.Whereabouts != nil {
+			whConfig := (*WhereaboutsT)(plugins.Whereabouts)
+			err := r.createPlugin(whConfig)
+			if err != nil {
+				return err
+			}
+			*fileList = append(*fileList, "whereabouts.yaml")
+		}
+
+		if plugins.HostPlumber != nil {
+			hostPlumberConfig := (*HostPlumberT)(plugins.HostPlumber)
+			err := r.createPlugin(hostPlumberConfig)
+			if err != nil {
+				return err
+			}
+			*fileList = append(*fileList, "hostplumber.yaml")
+		}
+
+		if plugins.NodeFeatureDiscovery != nil {
+			nfdConfig := (*NodeFeatureDiscoveryT)(plugins.NodeFeatureDiscovery)
+			err := r.createPlugin(nfdConfig)
+			if err != nil {
+				return err
+			}
+			*fileList = append(*fileList, "nfd.yaml")
+		}
+	}
+	return nil
+}
+
+func (r *NetworkPluginsReconciler) parseOldPlugins(fileList *[]string) error {
+	// First find out which plugins are missing from new spec vs old spec
+	if r.prevSpec == nil || r.prevSpec.CniPlugins == nil {
+		// Old spec was empty, nothing to delete
+		return nil
+	}
+
+	old := r.prevSpec.CniPlugins
+	os.MkdirAll(DeleteDir, os.ModePerm)
+
+	var noCni bool
+	if r.currentSpec.CniPlugins == nil {
+		noCni = true
+	} else {
+		noCni = false
+	}
+
+	if (noCni == true || r.currentSpec.CniPlugins.Multus == nil) && old.Multus != nil {
+		multusConfig := (*MultusT)(old.Multus)
+		err := r.deletePlugin(multusConfig)
+		if err != nil {
+			return err
+		}
+		*fileList = append(*fileList, "multus.yaml")
+	}
+
+	if (noCni == true || r.currentSpec.CniPlugins.Whereabouts == nil) && old.Whereabouts != nil {
+		whereaboutsConfig := (*WhereaboutsT)(old.Whereabouts)
+		err := r.deletePlugin(whereaboutsConfig)
+		if err != nil {
+			return err
+		}
+		*fileList = append(*fileList, "whereabouts.yaml")
+	}
+
+	if (noCni == true || r.currentSpec.CniPlugins.Sriov == nil) && old.Sriov != nil {
+		sriovConfig := (*SriovT)(old.Sriov)
+		err := r.deletePlugin(sriovConfig)
+		if err != nil {
+			return err
+		}
+		*fileList = append(*fileList, "sriov-cni.yaml")
+		*fileList = append(*fileList, "sriov-deviceplugin.yaml")
+	}
+
+	if (noCni == true || r.currentSpec.CniPlugins.HostPlumber == nil) && old.HostPlumber != nil {
+		hostPlumberConfig := (*HostPlumberT)(old.HostPlumber)
+		err := r.deletePlugin(hostPlumberConfig)
+		if err != nil {
+			return err
+		}
+		*fileList = append(*fileList, "hostplumber.yaml")
+	}
+
+	if (noCni == true || r.currentSpec.CniPlugins.NodeFeatureDiscovery == nil) && old.NodeFeatureDiscovery != nil {
+		nfdConfig := (*NodeFeatureDiscoveryT)(old.NodeFeatureDiscovery)
+		err := r.deletePlugin(nfdConfig)
+		if err != nil {
+			return err
+		}
+		*fileList = append(*fileList, "nfd.yaml")
+	}
+
+	return nil
+}
+
+func (r *NetworkPluginsReconciler) createOrDeletePlugins(networkPlugins *plumberv1.NetworkPlugins, fileList []string, mode string) error {
```

Comment:
Personal preference, but I would split this up into two (or three) functions `createPlugins`, `deletePlugins` and maybe a function that contains the common logic L479:L500 — as I find the `mode`-based toggling approach fragile and it seems like an indicator that the function could/should be split up. Though I have seen it more in our code... is it common in the OpenStack?